### ISSUE TITLE
swipe gesture for number pickers

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/gravity-force-lab-basics#242]
Currently, you can swipe up or down to incremement/decrement the values of the number spinner. Instead, we would like to try a continuous press and hold to drag operation. You should be able to double tap to drag, then drag up/right to continuously increase the value and down/left to continuously decrease the value.

**Comment by terracoda** [comment]
@jessegreenberg, thus far, we are trying to keep the instructions for grabbable draggables the same.

The general interaction pattern goes like this:
1. Swipe to focus
2. On focus, learner hears object name and current value of object (slider under the hood)
3. Learner double taps to get instructions
4. General instructions: “Tap and hold to grab and drag.” 
    - It’s nice to have a general pattern with general instructions, AND it might be important to have an option to create specific instructions for a particular interaction if needed. 
5. Tap and hold activates a grab and makes the object draggable. “Grabbed.” is announced.
6. While dragging, critical updates fire.
7. When finger is lifted, the object is released, and “Released.” is announced.

For the number spinners this custom swipe gesture would sound like this:
1. Swipe to mass control
2. On focus, I hear: 
    - Mass 1, 2 billion kilograms
3. Not sure what to do, so I double tap.
4. I get instructions (which are the same that I heard before on the spheres)
    - Tab and hold to grab and drag.
5. I execute my Tap and hold, and begin dragging right or up I hear:
6. Critical updates   
    - Grabbed.  3 billion kilograms + force changes, 
    - 4 billion kilograms + force changes
7. I lift my finger, and "Release" the mass control
    - "Released."

 We can try these default instructions, but I feel like the mass control is not an object that you think about dragging and grabbing a more physical object like a mass sphere, or a balloon, or even a hand or foot.

Different instructions for less object-like objects might be handy.

**Comment by terracoda** [comment]
@jessegreenberg, I have a feeling we are going to need some options for how we describe draggable slider interactions in a self-voicing context where a learner may or may not have vision.

**Comment by terracoda** [comment]
My specific concern is the very similar names and the identical instructions for the mass spheres and the mass controls.

**Comment by terracoda** [comment]
@jessegreenberg, ponderings and possible learner confusion aside, should we just implement the default situation (described above) and then see where we need to go from there.

I have a feeling, if we were to customize any instructions, it would be better to customize the mass sphere position instructions.

**Comment by jessegreenberg** [comment]
In the above commit I implemented the new gesture control so that press and hold -> drag will change the mass values.

I agree with all your above statements @terracoda, this could be confusing.

In the PDOM the labels are "Move Mass 1" for the sphere and "Mass 1" for the mass control. Were they given different labels to resolve this exact issue? Could we use different labels for self-voicing as well?

**Comment by terracoda** [comment]
Apologies @jessegreenberg, this was my number 1 issue for our meeting yesterday and when I clicked on the issue during the meeting, I thought it was something else that was resolved and we didn't discuss it. Apologies for my confusion.

One option, is indeed to change the labels.

In the context of the entire SVing content that is available, I want to remind myself that the learner has the option to turn on hints (help text strings). I wouldn't want to repeat stuff that could be delivered via hints.

In GFLB, we have 3 help-text/hints:
- Change position and mass of spheres to explore. (Initial hint, delivered from the hint button, and from focus on the Force reading block.)
- Move spheres, closer or farther from each other. (mass 1 and mass 2 spheres that change distance)
- Change mass of spheres in billions of kilograms. (Mass 1 and Mass 2 mass controls that change mass)

My ideas in the design doc for a different option for instructions for the number spinners was to focus the custom instructions needed for "sliders" on general interaction instructions that could be platform and interaction specific not sim-specific.

Ideas for Mass number spinners which conceptually are less "grabby" and more "draggy" in terms of objects:
Possible Instructions on double-tap
- Tap, hold and drag to increase or decrease.
- Tap and drag to increase or decrease.
- Tap and drag to adjust.
- Tap and drag to change.
**Then once activated with tap and hold, the object becomes grabbable or adjustable**
- Draggable
Object & context responses
Released.

**Comment by terracoda** [comment]
@jessegreenberg, I am wondering if would be easier to parameterize these instructions, for example:
- Tap and drag to {{change mass}}.

And would this pattern be more generalizable?
- Tap and drag to move sphere.
- Tap and drag to move hand.
- Tap and drag to swing leg/rub foot/move foot.
- Tap and hold to move balloon.
- Tap and hold to move ruler.
- Tab and hold to rub book.

**Comment by terracoda** [comment]
Might be related to https://github.com/phetsims/john-travoltage/issues/364

**Comment by terracoda** [comment]
- Tap, hold and drag to move sphere.
- Tap, hold and drag to move hand.
- Tap, hold and drag to swing leg/rub foot/move foot.
- Tap, hold and drag to move balloon.
- Tap, hold and drag to move ruler.
- Tap, hold and drag to rub book.

**Comment by terracoda** [comment]
See also John Travoltage issue https://github.com/phetsims/john-travoltage/issues/364

**Comment by terracoda** [comment]
**Changes for the sphere sliders**
- Tap, hold, and drag to move mass 1, blue sphere
Grabbed/ Released

- Tap, hold, and drag to move mass 2, red sphere
Grabbed/ Released

**Changes for the mass number spinners**
- Tap, hold, and drag to change mass
Draggable/ Released

- Tap, hold, and drag to change mass
Draggable/ Released

**Comment by jessegreenberg** [comment]
Done in the above commit.

**Comment by terracoda** [comment]
@jessegreenberg, re-opening this issue because in the latest test flight, I am hearing neither:
- the new gesture instruction nor,
- the new "Draggable" response for when a mass control is activated for dragging.

I am assuming since "Draggable" might be a totally new thing, and you might still be working on it. Possibly, even in a different issue?

The mass spheres are fine. It is the mass control number spinners that still have the original instructions and "Grabbed" response.

My understanding is that they would both be updated to the same hint, "to change mass":

**Mass 1**
- Tap, hold, and drag to change mass
- Draggable/ Released

**Mass 2**
- Tap, hold, and drag to change mass
- Draggable/ Released

**Comment by jessegreenberg** [comment]
Thanks @terracoda, sorry about that. Just totally forgot to change the content for the masses controls.

**Comment by jessegreenberg** [comment]
OK, change to mass controls made in above commit. Keeping open for review, hopefully during meeting today.

**Comment by terracoda** [comment]
Discussed at meeting.
We will change the general hint to 2 different hints:
**Mass 1**
- Tap, hold, and drag to change mass of Mass 1.
- Draggable/ Released

**Mass 2**
- Tap, hold, and drag to change mass of Mass 2.
- Draggable/ Released

**Comment by jessegreenberg** [comment]
Change made in above commit, ready for review @terracoda.

**Comment by terracoda** [comment]
@jessegreenberg, the descriptions of the gesture interaction for the mass number spinner is correctly implemented. and sounds good to me. Thanks!

Closing this issue.
